engine/runtime/src/main/java/org/enso/interpreter/runtime/ResourceManager.java

```diff
@@ -174,21 +147,82 @@ public void shutdown() {
   }
 
   /**
-   * The worker action for the underlying logic of this module. At least one such thread must be
-   * spawned in order for this module to be operational.
+   * Processes {@link Item}s eligible for GC. Plays two roles. First of all cleans {@link
+   * #referenceQueue} in {@link #run()} method running in its own thread. Then it invokes finalizers
+   * in {@link #perform} method inside of Enso execution context.
    */
-  private class Runner implements Runnable {
+  private final class ProcessItems extends ThreadLocalAction implements Runnable {
+    /**
+     * @GuardedBy("pendingItems")
+     */
+    private final List<Item> pendingItems = new ArrayList<>();
+
+    /**
+     * @GuardedBy("pendingItems")
+     */
+    private Future<Void> request;
+
     private volatile boolean killed = false;
 
+    ProcessItems() {
+      super(false, false, true);
+    }
+
+    /**
+     * Runs at a safe point in middle of regular Enso program execution. Gathers all available
+     * {@link #pendingItems} and runs their finalizers. Removes all processed items from {@link
+     * #pendingItems}. If there are any remaining, continues processing them. Otherwise finishes.
+     *
+     * @param access not used for anything
+     */
+    @Override
+    protected void perform(ThreadLocalAction.Access access) {
+      for (; ; ) {
+        Item[] toProcess;
+        synchronized (pendingItems) {
+          request.cancel(false);
```

Comment:
`request` is guaranteed to be non-null at this point?

Reply:
It should be non-`null`. To get into this `process` method, a call to [submitThreadLocal](https://github.com/enso-org/enso/pull/10602/files#diff-8afe630a7fa9f76767a43dadee9ce42219dce2d752c325909006cc527809c178R222) must be made and it assigns the `request`.

The `request` is assigned to `null` in this method, just before `return`.

E.g. unless there is some re-entrant invocation of the `process` method (it was there, but I hopefully fixed it), `request` shall not be `null` at this point.

Reply:
Shall we maybe include an assert at least?

Reply:
What's a difference between `NullPointerException` and `AssertionError`?